### PR TITLE
Alternative scoll/scan implementation

### PR DESCRIPTION
**Pull request by beowulfenator**
Hi!

I've implemented my own version of batch query for elasticsearch. It is consistent with other db classes. In other words, batch query processing is done with `batch()` and `each()` functions. Also, I've written some tests for this functionality.

I would love to see this commit merged into trunk, because then I shall be able to fix issues #48 and #19.

### CODE REVIEW

**Comment by cebe** [comment]
This looks awesome, thanks! Will make sure this goes into the next release.

**Comment by beowulfenator** [comment]
Should I fix `updateAll` and `deleteAll` in this branch as well?

**Comment by cebe** [comment]
not sure how this is related? you mean to use the scroll api to iterate over the models to update/delete?

**Comment by beowulfenator** [comment]
Exactly! The problem with updateAll/deleteAll is that it first needs to get all document ids, then perform the action. As far as I know, search API doesn't even let you get all results (you can set `size` to max_int or some really high value, but that's a bad solution). Even if it did, it would be terribly inefficient for anything over a few hundred records. So I suggest using scroll api in scan mode to get all ids first, and then carry on as usual.

**Comment by cebe** [comment]
Good idea, afair, setting size to max_int will result in ES to allocate that amount of memory regardsless of the number of results or so. May not be the case anymore but using scroll api sounds like a good idea.

**Comment by beowulfenator** [comment]
My last commit fixes issues #48 and #39.

File: BatchQueryResult.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+     */
+    private $_batch = null;
+    /**
+     * @var mixed the value for the current iteration
+     */
+    private $_value = null;
+    /**
+     * @var string|integer the key for the current iteration
+     */
+    private $_key = null;
```
Comment (cebe) [inline BatchQueryResult.php]: parent method key() and current() will not access the private members in this class but the ones of the parent class.
I'd prefer to duplicate the BatchQueryResult from framework here, we may later extract a proper base class from these if it makes sense.

File: Command.php [hunk 2]
```diff
@@ -320,6 +320,26 @@ public function closeIndex($index)
     }
 
     /**
+     * @param array $options
+     * @see https://www.elastic.co/guide/en/elasticsearch/guide/current/scan-scroll.html
```
Comment (cebe) [inline Command.php]: url shows 404 :-/

**Comment by cebe** [comment]
looks good so far except for the comments added. Also needs a CHANGELOG entry and `@since 2.0.4` annotations.

**Comment by cebe** [comment]
:+1:

**Pull request merged**